### ISSUE TITLE
pandas version

### ISSUE THREAD

**Issue by ews-ffarella** [NREL/OpenOA#281]
Thanks for the great lib!

It seems that pandas 2.2 breaks `test/regression/electrical_losses.py`. Pinning to pandas<2.2 works for me!

**Comment by RHammond2** [comment]
@ews-ffarella thanks for flagging this! I was just testing this out on my own machine and both Pandas 2.2.0 and 2.2.1 work for me, would you be able to share the full traceback of the error message? If this is fairly straightforward it would be great to include a patch in our upcoming 3.1 release.

**Comment by ews-ffarella** [comment]
Hi!
I think it has to do with datetime handling.

`pip list`

```bash
Package                       Version         Editable project location
----------------------------- --------------- -------------------------------------
accessible-pygments           0.0.4
alabaster                     0.7.16
anyio                         4.3.0
argon2-cffi                   23.1.0
argon2-cffi-bindings          21.2.0
arrow                         1.3.0
asttokens                     2.4.1
async-lru                     2.0.4
attrs                         23.2.0
Babel                         2.14.0
beautifulsoup4                4.12.3
black                         24.2.0
bleach                        6.1.0
bokeh                         3.3.4
cdsapi                        0.6.1
certifi                       2024.2.2
cffi                          1.16.0
cfgv                          3.4.0
cftime                        1.6.3
charset-normalizer            3.3.2
click                         8.1.7
cloudpickle                   3.0.0
comm                          0.2.1
contourpy                     1.2.0
coverage                      7.4.3
cycler                        0.12.1
dask                          2024.2.1
debugpy                       1.8.1
decorator                     5.1.1
defusedxml                    0.7.1
distlib                       0.3.8
docutils                      0.20.1
eia-python                    1.22
exceptiongroup                1.2.0
executing                     2.0.1
fastjsonschema                2.19.1
filelock                      3.13.1
flake8                        7.0.0
flake8-docstrings             1.7.0
fonttools                     4.49.0
fqdn                          1.5.1
fsspec                        2024.2.0
greenlet                      3.0.3
h11                           0.14.0
h5pyd                         0.18.0
httpcore                      1.0.4
httpx                         0.27.0
identify                      2.5.35
idna                          3.6
imagesize                     1.4.1
importlib-metadata            7.0.1
iniconfig                     2.0.0
ipykernel                     6.29.3
ipython                       8.22.2
isoduration                   20.11.0
isort                         5.13.2
jedi                          0.19.1
Jinja2                        3.1.3
joblib                        1.3.2
json5                         0.9.20
jsonpointer                   2.4
jsonschema                    4.21.1
jsonschema-specifications     2023.12.1
jupyter-cache                 1.0.0
jupyter_client                8.6.0
jupyter_core                  5.7.1
jupyter-events                0.9.0
jupyter-lsp                   2.2.4
jupyter_server                2.13.0
jupyter_server_terminals      0.5.2
jupyterlab                    4.1.3
jupyterlab_pygments           0.3.0
jupyterlab_server             2.25.3
kiwisolver                    1.4.5
latexcodec                    3.0.0
locket                        1.0.0
markdown-it-py                3.0.0
MarkupSafe                    2.1.5
matplotlib                    3.8.3
matplotlib-inline             0.1.6
mccabe                        0.7.0
mdit-py-plugins               0.4.0
mdurl                         0.1.2
mistune                       3.0.2
mypy-extensions               1.0.0
myst-nb                       1.0.0
myst-parser                   2.0.0
nbclient                      0.9.0
nbconvert                     7.16.2
nbformat                      5.9.2
nest-asyncio                  1.6.0
netCDF4                       1.6.5
nodeenv                       1.8.0
notebook_shim                 0.2.4
numpy                         1.26.4
OpenOA                        3.0.1           /home/ffarella/EWS/third-party/OpenOA
overrides                     7.7.0
packaging                     23.2
pandas                        2.2.1
pandocfilters                 1.5.1
parso                         0.8.3
partd                         1.4.1
pathspec                      0.12.1
patsy                         0.5.6
pexpect                       4.9.0
pillow                        10.2.0
pip                           24.0
platformdirs                  4.2.0
pluggy                        1.4.0
pre-commit                    3.6.2
progressbar2                  4.4.2
prometheus_client             0.20.0
prompt-toolkit                3.0.43
psutil                        5.9.8
ptyprocess                    0.7.0
pure-eval                     0.2.2
py4j                          0.10.9.7
pybtex                        0.24.0
pybtex-docutils               1.0.3
pycodestyle                   2.11.1
pycparser                     2.21
pydata-sphinx-theme           0.15.2
pydocstyle                    6.3.0
pyflakes                      3.2.0
pygam                         0.9.1
Pygments                      2.17.2
PyJWT                         2.8.0
pyparsing                     3.1.2
pyproj                        3.6.1
pyspark                       3.5.1
pytest                        8.0.2
pytest-cov                    4.1.0
python-dateutil               2.9.0.post0
python-json-logger            2.0.7
python-utils                  3.8.2
pytz                          2024.1
PyYAML                        6.0.1
pyzmq                         25.1.2
referencing                   0.33.0
requests                      2.31.0
requests-unixsocket           0.3.0
rfc3339-validator             0.1.4
rfc3986-validator             0.1.1
rpds-py                       0.18.0
scikit-learn                  1.4.1.post1
scipy                         1.11.4
Send2Trash                    1.8.2
setuptools                    69.1.0
shapely                       2.0.3
six                           1.16.0
sniffio                       1.3.1
snowballstemmer               2.2.0
soupsieve                     2.5
Sphinx                        7.2.6
sphinx_design                 0.5.0
sphinxcontrib-applehelp       1.0.8
sphinxcontrib-bibtex          2.6.2
sphinxcontrib-devhelp         1.0.6
sphinxcontrib-htmlhelp        2.0.5
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.7
sphinxcontrib-serializinghtml 1.1.10
SQLAlchemy                    2.0.28
stack-data                    0.6.3
statsmodels                   0.14.1
tabulate                      0.9.0
terminado                     0.18.0
threadpoolctl                 3.3.0
tinycss2                      1.2.1
tomli                         2.0.1
toolz                         0.12.1
tornado                       6.4
tqdm                          4.66.2
traitlets                     5.14.1
types-python-dateutil         2.8.19.20240106
typing_extensions             4.10.0
tzdata                        2024.1
uri-template                  1.3.0
urllib3                       2.2.1
virtualenv                    20.25.1
wcwidth                       0.2.13
webcolors                     1.13
webencodings                  0.5.1
websocket-client              1.7.0
wheel                         0.42.0
xarray                        2024.2.0
xyzservices                   2023.10.1
zipp                          3.17.0
```

```bash
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-8.0.2, pluggy-1.4.0 -- /home/ffarella/EWS/third-party/OpenOA/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/ffarella/EWS/third-party/OpenOA
configfile: pytest.ini
plugins: anyio-4.3.0, cov-4.1.0
collecting ... collected 2 items

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results <IPython.core.display.Markdown object>
FAILED
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results <IPython.core.display.Markdown object>
FAILED

=================================== FAILURES ===================================
______________ TestElectricalLosses.testelectrical_losses_results ______________

self = <electrical_losses.TestElectricalLosses testMethod=testelectrical_losses_results>

    def setUp(self):
        np.random.seed(42)
        # Set up data to use for testing (ENGIE data)
        self.project = project_ENGIE.prepare(example_data_path_str, use_cleansed=False)
        self.project.analysis_type.append("ElectricalLosses")
>       self.project.validate()

test/regression/electrical_losses.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
openoa/logging.py:29: in _wrapper
    return the_method(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = None, metadata = None

    @logged_method_call
    def validate(self, metadata: Optional[dict | str | Path | PlantMetaData] = None) -> None:
        """Secondary method to validate the plant data objects after loading or changing
        data with option to provide an updated `metadata` object/file as well
    
        Args:
            metadata (Optional[dict]): Updated metadata object, dictionary, or file to
                create the updated metadata for data validation, which should align with
                the mapped column names during initialization.
    
        Raises:
            ValueError: Raised at the end if errors are caught in the validation steps.
        """
        logger.info("Post-intialization data validation")
        # Put the index columns back into the column space to ensure success of re-validation
        self._unset_index_columns()
    
        # Initialization will have converted the column naming convention, but an updated
        # metadata should account for the renaming of the columns
        if metadata is None:
            self.update_column_names(to_original=True)
        else:
            self.metadata = metadata
    
        # Reset the index columns to be part of the columns space so the validations still work
        self._errors = {
            "missing": self._validate_column_names(),
            "dtype": self._validate_dtypes(),
        }
    
        self._set_index_columns()
        self._errors["frequency"] = self._validate_frequency()
    
        error_message = _compose_error_message(self._errors, self.metadata, self.analysis_type)
        if error_message:
>           raise ValueError(error_message)
E           ValueError: `reanalysis-era5` data is of the wrong frequency: h
E           `reanalysis-merra2` data is of the wrong frequency: h

openoa/plant.py:1024: ValueError
_____________ TestElectricalLossesUQ.testelectrical_losses_results _____________

self = <electrical_losses.TestElectricalLossesUQ testMethod=testelectrical_losses_results>

    def setUp(self):
        np.random.seed(42)
        # Set up data to use for testing (ENGIE data)
        self.project = project_ENGIE.prepare(example_data_path_str, use_cleansed=False)
        self.project.analysis_type.append("ElectricalLosses")
>       self.project.validate()

test/regression/electrical_losses.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
openoa/logging.py:29: in _wrapper
    return the_method(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = None, metadata = None

    @logged_method_call
    def validate(self, metadata: Optional[dict | str | Path | PlantMetaData] = None) -> None:
        """Secondary method to validate the plant data objects after loading or changing
        data with option to provide an updated `metadata` object/file as well
    
        Args:
            metadata (Optional[dict]): Updated metadata object, dictionary, or file to
                create the updated metadata for data validation, which should align with
                the mapped column names during initialization.
    
        Raises:
            ValueError: Raised at the end if errors are caught in the validation steps.
        """
        logger.info("Post-intialization data validation")
        # Put the index columns back into the column space to ensure success of re-validation
        self._unset_index_columns()
    
        # Initialization will have converted the column naming convention, but an updated
        # metadata should account for the renaming of the columns
        if metadata is None:
            self.update_column_names(to_original=True)
        else:
            self.metadata = metadata
    
        # Reset the index columns to be part of the columns space so the validations still work
        self._errors = {
            "missing": self._validate_column_names(),
            "dtype": self._validate_dtypes(),
        }
    
        self._set_index_columns()
        self._errors["frequency"] = self._validate_frequency()
    
        error_message = _compose_error_message(self._errors, self.metadata, self.analysis_type)
        if error_message:
>           raise ValueError(error_message)
E           ValueError: `reanalysis-era5` data is of the wrong frequency: h
E           `reanalysis-merra2` data is of the wrong frequency: h

openoa/plant.py:1024: ValueError
=============================== warnings summary ===============================
test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/examples/project_ENGIE.py:106: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
  You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
  A typical example is when you are setting values in a column of a DataFrame, like:
  
  df["col"][row_indexer] = value
  
  Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.
  
  See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
    scada_df.loc[ix_turbine, "Ot_avg"].loc[ix_flag.values.flatten()] = np.nan

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/unit_conversion.py:32: FutureWarning: 'T' is deprecated and will be removed in a future version, please use 'min' instead.
    _dt_range = pd.date_range(start="09/30/2022", periods=2, freq=sample_rate_min)

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/examples/project_ENGIE.py:258: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
    reanalysis_era5_df = reanalysis_era5_df.asfreq("1H")

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'S' is deprecated and will be removed in a future version. Please use 's' instead of 'S'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'T' is deprecated and will be removed in a future version. Please use 'min' instead of 'T'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'N' is deprecated and will be removed in a future version. Please use 'ns' instead of 'N'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'H' is deprecated and will be removed in a future version. Please use 'h' instead of 'H'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'U' is deprecated and will be removed in a future version. Please use 'us' instead of 'U'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
  /home/ffarella/EWS/third-party/OpenOA/openoa/utils/timeseries.py:31: FutureWarning: 'L' is deprecated and will be removed in a future version. Please use 'ms' instead of 'L'.
    seconds = pd.to_timedelta(f"1{offset}").total_seconds()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED test/regression/electrical_losses.py::TestElectricalLosses::testelectrical_losses_results
FAILED test/regression/electrical_losses.py::TestElectricalLossesUQ::testelectrical_losses_results
======================== 2 failed, 26 warnings in 7.33s ========================

```

**Comment by RHammond2** [comment]
Which branch are you using @ews-ffarella? On my main branch many tests fail with pandas 2.2, and on develop I'm not able to reproduce your results. If you're on develop, would you be able to confirm that it's up to date with the latest changes? I just want to make sure we're on the same page here for diagnosing the issue.

**Comment by ews-ffarella** [comment]
> Which branch are you using @ews-ffarella? On my main branch many tests fail with pandas 2.2, and on develop I'm not able to reproduce your results. If you're on develop, would you be able to confirm that it's up to date with the latest changes? I just want to make sure we're on the same page here for diagnosing the issue.

I was on the main branch. Sorry about that. With 2.1.4, everything is fine. 

This is great stuff by the way.  
Should I post my "issues" suggestion all in separate issues, or only in a single one?

There are not many though. The only thing that tripped me is that one had to make sure the the scada data and the reanalyis start at the same time. I had a 10min gap and it lead to a lot of NaNs in the daily aggregate df. I can publish my branch if you want.

**Comment by RHammond2** [comment]
No worries, we were expecting to have the v3.1 release a bit sooner given this exact issue you're encountering!

In general, I'd say to checkout our [issue submission checklist](https://openoa.readthedocs.io/en/latest/getting_started/contributing.html#issue-tracking), but I think highly related/overlapping topics can go under a single issue.

It'd be great to learn more about the gap in starting timestamps, especially so we could find a way to fix that in OpenOA. Feel free to open an issue for that specifically with a link to your branch, and provide as much as you can so we can diagnosis, or even reproduce the issue.

For now I'll mark this as informally closed, but hold off until 3.1 is released since the issue still exists in the main branch.
